### PR TITLE
C++ macro expander

### PR DESCRIPTION
**Pull request by dwightguth**
This adds a new tool to the llvm backend, kore-expand-macros, which takes a kompiled directory and a pattern and returns the pattern with all macros expanded. It also will reverse expand alias macros when you invoke `kprint`.

### CODE REVIEW

File: lib/ast/AST.cpp
Comment (virgil-serbanuta) [inline lib/ast/AST.cpp:980]: Why does this case "fall through" and get resolved as argument matching below? I.e., why doesn't this just return `false`, or why doesn't `inj` get decomposed as in the other subsorting case?

Is it because the only case where it would match would be when `this` is an `inj(variable:b)`, and it's ok to have a substitution from `variable` to `stuff:a` without having an inj on top of it, i.e. inj(a, b)(stuff)?

File: lib/ast/AST.cpp
Comment (dwightguth) [reply lib/ast/AST.cpp:948]: appliedRules is a reference so yes, this is necessary.

File: include/kllvm/ast/AST.h
Comment (dwightguth) [reply include/kllvm/ast/AST.h:709]: No particular reason. The code is better when written iteratively, so I went ahead and changed it, but note that this code was just copy pasted and turned into a template function in this PR; the actual algorithm wasn't changed at all in this diff before now.

File: lib/ast/AST.cpp
Comment (dwightguth) [reply lib/ast/AST.cpp:980]: This code was based originally on the code for pattern matching in the llvm backend, and originally this case worked the way it did because of the way that algorithm worked, but I think I mistranslated it to the new domain. I'm going to go ahead and do what you suggest and decompose it via the subsort.

**Review by theo25 — APPROVED** [review]
I went through all the changes. I didn't notice anything bad, but I am not fully up-to-date with all the code base at this point so I may have missed something. I would wait for the rest of the reviews before merging.

File: lib/ast/AST.cpp
Comment (virgil-serbanuta) [reply lib/ast/AST.cpp:948]: Ok, I actually meant this: Is the referred value used anywhere after the `rhs->substitute(subst)->expandMacros` call? As I see it:
* KorePattern::expandMacros with 4 arguments passes a reference to a local object when it adds `appliedRules` as an extra argument. It never uses it after the call.
* `KORECompositePattern::expandMacros` with 5 arguments uses it only inside an `if` statement that calls `expandMacros` with a changed `appliedRules`, then returns from the function, so it does not use this value after the call.
* I think that the above are the only calls where `appliedRules` is used as an argument, so I would say that there is no way of using it after the call.

Assuming that's correct, then, if you still want to restore the old value, then how about this: you could make `appliedRules` be a const reference, then you could rewrite the `if` like this:
```c
if (stuff) {
￼      std::set<size_t> newAppliedRules = appliedRules;
￼      newAppliedRules.insert(i);
￼      return rhs->substitute(subst)->expandMacros(subsorts, overloads, macros, reverse, newAppliedRules);
}
```

File: lib/ast/AST.cpp
Comment (virgil-serbanuta) [inline lib/ast/AST.cpp:993]: Does it make sense to also check for this overloads stuff in `this`, i.e. not only in `subj`?

File: lib/ast/AST.cpp
Comment (dwightguth) [reply lib/ast/AST.cpp:948]: I can tell you that it IS used after the call because it's used after the function returns. I can't make it a const reference because it's actually modified. Regardless, this code is simply an accurate translation of some Java code that this entire algorithm is based on, and I'm not going to change it to be less like the original just because it is allegedly stylistically better, because it's not at all clear to me that the change you are proposing preserves behavior. And even if it in fact does, it's better that the code resemble the original as much as possible.

File: lib/ast/AST.cpp
Comment (dwightguth) [reply lib/ast/AST.cpp:993]: This is an accurate representation of how matching works in the llvm backend. I don't think it's appropriate to explore changed behavior here.

**Review by virgil-serbanuta — APPROVED** [review]

**Pull request merged**
